### PR TITLE
Migrate some duplicate binding tests to traverse

### PR DESCRIPTION
**Pull request by danez**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | `Fixes #1, Fixes #2` 
| Patch: Bug Fix?          | y
| Major: Breaking Change?  | ?
| Minor: New Feature?      | n
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

The main goal here was to transfer all the tests which test duplicate bindings. They were in different transforms, although the code was in `@babel/traverse` and the transforms have no impact on that behavior, and imho also shouldn't. 

While doing this I discovered, that on case wasn't throwing an error:

```js
let foo;
function foo(){};
```

see inline comment.

### CODE REVIEW

File: packages/babel-traverse/src/scope/index.js [hunk 1]
```diff
@@ -349,9 +349,6 @@ export default class Scope {
     // class expression
     if (local.kind === "local") return;
 
-    // ignore hoisted functions if there's also a local let
-    if (kind === "hoisted" && local.kind === "let") return;
-
```
Comment (danez) [inline packages/babel-traverse/src/scope/index.js]: For some reason we have this exception in the code for the case mentioned in the PR. I couldn't come up with a reason why we do this and the commit that introduced it doesn't help either 8a4296a948b496e9fde1d6271a4e0dbdeaacb128
Simply removing it also doesn't break any of the tests. 
Do you think it is safe to remove? Anyone has an idea why this code was there and which case it covers?

Comment (nicolo-ribaudo) [reply packages/babel-traverse/src/scope/index.js]: Maybe it was this? It is an Annex B thing:
```js
let a;
{ function a() {} }
```

File: packages/babel-traverse/test/scope.js [hunk 2]
```diff
@@ -250,4 +271,84 @@ describe("scope", function() {
...
+          const ast = createTryCatch(name);
+
+          expect(() => getPath(ast)).toThrowErrorMatchingSnapshot();
+        });
+      });
+
+      it("var", () => {
+        const ast = createTryCatch("var");
+
+        expect(getPath(ast).node).toMatchSnapshot();
```
Comment (nicolo-ribaudo) [inline packages/babel-traverse/test/scope.js]: Nit: `generator(getPath(ast).node)` would make the snapshot output easier to check.

File: packages/babel-traverse/src/scope/index.js (hunk 1 above)
Comment (danez) [reply packages/babel-traverse/src/scope/index.js]: right but even with removing these lines, this case still works. will add a test.

File: packages/babel-traverse/test/scope.js (hunk 2 above)
Comment (danez) [reply packages/babel-traverse/test/scope.js]: I was thinking exactly the same, I guess what I'm really testing here is that it does not throw. So `expect().not.toThrow()` makes sense to me.

**Comment by nicolo-ribaudo** [comment]
Maybe it's wort using `.toThrowErrorMatchingInlineSnapshot` instead of `toThrowErrorMatchingSnapshot`, since the error messages are small?

**Comment by danez** [comment]
Oh I have never seen this. makes sense.

**Comment by danez** [comment]
Nevermind this does not work obviously with loops and generated testcases.

File: packages/babel-traverse/test/scope.js [hunk 3]
```diff
@@ -250,4 +271,103 @@ describe("scope", function() {
...
+            ),
+          ]),
+        ];
+
+        expect(() => getPath(ast)).not.toThrow();
+      });
+    });
+
+    describe("global", () => {
+      ["let", "const"].forEach(name => {
```
Comment (nicolo-ribaudo) [inline packages/babel-traverse/test/scope.js]: We should also test `class/function` and `class/var` (and maybe `function/var` to be sure that it doesn't throw)

Comment (danez) [reply packages/babel-traverse/test/scope.js]: Okay I added a more generic way to define tests.
Turns out we have some problems, unless I made a mistake in the tests.
```
var foo = 1;
class foo {};
```
and
```
class foo {};
var foo = 1;
```
Class and var never throw.

also does not throw:

```js
function foo() {}
const foo = 1;
```

```js
var foo = 1;
const foo = 1;
```

Thats why I disabled the reverse matching in the test.

**Comment by danez** [comment]
It seems the scope tracking in traverse has some flaws. I guess we should have a look separately from this. I can maybe work on this once done with the scope tracking in the parser.
I was also already thinking that maybe we can make a shared module that handles scopes correctly and share it between the parser and traverse. That would be nice, but not sure what requirements the scope tracking in traverse has.

**Comment by nicolo-ribaudo** [comment]
Yeah let's just update the tests for now.

> I was also already thinking that maybe we can make a shared module that handles scopes correctly and share it between the parser and traverse

traverse's scope tracker needs to be kept updated while parser's scope describes a static AST; I don't know how feasible it would be.

**Comment by danez** [comment]
> traverse's scope tracker needs to be kept updated while parser's scope describes a static AST; I don't know how feasible it would be.

But in the end both implementations (parser and traverse) work on nodes, have a stack of scopes, add/remove bindings and check for duplicates. But yes there might be more stuff going on in traverse which I'm not yet aware of.

**Pull request merged**